cmd/pj-rehearse/main_test.go

```diff
@@ -0,0 +1,103 @@
+package main
+
+import (
+	"path/filepath"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	clientgo_testing "k8s.io/client-go/testing"
+	"k8s.io/test-infra/prow/apis/prowjobs/v1"
+	"k8s.io/test-infra/prow/client/clientset/versioned/fake"
+	prowconfig "k8s.io/test-infra/prow/config"
+	prowgithub "k8s.io/test-infra/prow/github"
+	"k8s.io/test-infra/prow/kube"
+	pjdwapi "k8s.io/test-infra/prow/pod-utils/downwardapi"
+
+	"github.com/openshift/ci-operator-prowgen/pkg/diffs"
+	"github.com/openshift/ci-operator-prowgen/pkg/rehearse"
+)
+
+var (
+	candidatePath = "../../test/pj-rehearse/candidate"
+	masterPath    = "../../test/pj-rehearse/master"
+)
+
+func TestIntergration(t *testing.T) {
+	jobSpec := pjdwapi.JobSpec{
+		Type:      kube.PresubmitJob,
+		Job:       "pull-pj-rehearse-demo",
+		BuildID:   "0",
+		ProwJobID: "prowjob",
+		Refs: &kube.Refs{
+			Org:     "super",
+			Repo:    "duper",
+			BaseRef: "base-ref",
+			BaseSHA: "base-sha",
+			Pulls: []v1.Pull{{
+				Number: 2596,
+				Author: "dptp-team",
+				SHA:    "a4dc6278a189aa66ae9c2a0f11834499f104fcd4",
+			}},
+		},
+	}
+
+	prFields := logrus.Fields{prowgithub.OrgLogField: jobSpec.Refs.Org, prowgithub.RepoLogField: jobSpec.Refs.Repo}
+	logger := logrus.WithFields(prFields)
+
+	candidateConfigPath := filepath.Join(candidatePath, diffs.ConfigInRepoPath)
+	candidateJobConfigPath := filepath.Join(candidatePath, diffs.JobConfigInRepoPath)
+	masterConfigPath := filepath.Join(masterPath, diffs.ConfigInRepoPath)
+	masterJobConfigPath := filepath.Join(masterPath, diffs.JobConfigInRepoPath)
+
+	prowConfig, err := prowconfig.Load(masterConfigPath, masterJobConfigPath)
+	if err != nil {
+		t.Fatalf("failed to load Prow config: %v", err)
+	}
+
+	prowPRConfig, err := prowconfig.Load(candidateConfigPath, candidateJobConfigPath)
+	if err != nil {
+		t.Fatalf("failed to load Prow config: %v", err)
+	}
+
+	fakecs := fake.NewSimpleClientset()
+	fakeclient := fakecs.ProwV1().ProwJobs("test-namespace")
+
+	changedPresubmits, err := diffs.GetChangedPresubmits(prowConfig, prowPRConfig)
+	if err != nil {
+		t.Fatalf("Failed to determine which jobs should be rehearsed: %v", err)
+	}
+
+	pjStatusChanger(fakecs, changedPresubmits, jobSpec.Refs, t)
+	if err := rehearse.ExecuteJobs(changedPresubmits, jobSpec.Refs.Pulls[0].Number, candidatePath, jobSpec.Refs, true, logger, fakeclient); err != nil {
+		t.Fatalf("Failed to execute rehearsal jobs: %v", err)
+	}
+}
+
+func pjStatusChanger(fakecs *fake.Clientset, toBeRehearsed map[string][]prowconfig.Presubmit, refs *v1.Refs, t *testing.T) error {
+	fakeclient := fakecs.ProwV1().ProwJobs("test-namespace")
+
+	watcher, err := fakeclient.Watch(metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Failed to setup watch: %v", err)
+	}
+
+	fakecs.Fake.PrependWatchReactor("prowjobs", func(clientgo_testing.Action) (bool, watch.Interface, error) {
```

Comment:
I think this import is most commonly `clienttesting`

Reply:
I don't really understand what this is doing -- what's the intent here?

Reply:
It will simulate a status change of the prowjob. Otherwise, `ExecuteJobs` will just hang forever.

Reply:
asking in a state of ignorace, Is there any better way to achieve this?